IdentityCore/src/throttling/cache/MSIDLRUCache.m

```diff
@@ -0,0 +1,432 @@
+//
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.  
+
+
+#import <Foundation/Foundation.h>
+#import "MSIDLRUCache.h"
+
+static NSString *const HEAD_SIGNATURE = @"HEAD";
+static NSString *const TAIL_SIGNATURE = @"TAIL";
+
+#define DEFAULT_CACHE_SIZE 1000
+#define DEFAULT_SIGNATURE_LENGTH 8
+
+//Helper class
+@interface MSIDLRUCacheNode : NSObject
+
+@property (nonatomic, readonly) NSString *signature;
+@property (nonatomic) NSMutableString *prevSignature;
+@property (nonatomic) NSMutableString *nextSignature;
+@property (nonatomic) id cacheRecord;
+
+- (instancetype)initWithSignature:(NSString *)signature
+                    prevSignature:(NSString *)prevSignature
+                    nextSignature:(NSString *)nextSignature
+                      cacheRecord:(id)cacheRecord;
+
+@end
+
+
+@implementation MSIDLRUCacheNode
+
+- (instancetype)initWithSignature:(NSString *)signature
+                    prevSignature:(NSString *)prevSignature
+                    nextSignature:(NSString *)nextSignature
+                      cacheRecord:(id)cacheRecord
+{
+    self = [super init];
+    if (self)
+    {
+        _signature = signature;
+        _prevSignature = [prevSignature mutableCopy];
+        _nextSignature = [nextSignature mutableCopy];
+        _cacheRecord = cacheRecord;
+    }
+    return self;
+}
+
+@end
```

Comment:
Since these are private helper classes not used by anyone else, I thought it best to keep it in the same file. Also, by doing so, I can avoid having mutable public properties.